File: SIN BDD/RESTFULL_JAVA_CONUNI_GR06/05. CLIESCRITORIO/cliente_escritorio_restful_java_conuni_gr06/src/ec/edu/monster/vista/PanelConversion.java
package ec.edu.monster.vista;

import ec.edu.monster.modelo.Resultado;
import java.awt.Color;
import java.util.Arrays;
import java.util.List;
import java.util.function.BiConsumer;
import javax.swing.DefaultComboBoxModel;

/**
 * Panel de conversion en formato Matisse (archivo .form emparejado).
 * Editable en NetBeans -> Design view.
 *
 * Es parametrizable: llamar {@link #setCategoria(String)} con "longitud",
 * "masa" o "temperatura" cambia el titulo y las opciones del combo.
 */
public class PanelConversion extends javax.swing.JPanel {

    public static class Opcion {
        public final String clave;
        public final String etiqueta;
        public Opcion(String clave, String etiqueta) {
            this.clave = clave;
            this.etiqueta = etiqueta;
        }
        @Override public String toString() { return etiqueta; }
    }

    private static final List<Opcion> OPCIONES_LONGITUD = Arrays.asList(
        new Opcion("metrosAPies",          "Metros a Pies"),
        new Opcion("kilometrosAMillas",    "Kilómetros a Millas"),
        new Opcion("centimetrosAPulgadas", "Centímetros a Pulgadas"),
        new Opcion("yardasAMetros",        "Yardas a Metros"),
        new Opcion("milimetrosAPulgadas",  "Milímetros a Pulgadas")
    );
    private static final List<Opcion> OPCIONES_MASA = Arrays.asList(
        new Opcion("kilogramosALibras",    "Kilogramos a Libras"),
        new Opcion("gramosAOnzas",         "Gramos a Onzas"),
        new Opcion("toneladasAKilogramos", "Toneladas a Kilogramos"),
        new Opcion("librasAOnzas",         "Libras a Onzas"),
        new Opcion("miligramosAGramos",    "Miligramos a Gramos")
    );
    private static final List<Opcion> OPCIONES_TEMPERATURA = Arrays.asList(
        new Opcion("celsiusAFahrenheit", "Celsius a Fahrenheit"),
        new Opcion("fahrenheitACelsius", "Fahrenheit a Celsius"),
        new Opcion("celsiusAKelvin",     "Celsius a Kelvin"),
        new Opcion("kelvinACelsius",     "Kelvin a Celsius"),
        new Opcion("fahrenheitAKelvin",  "Fahrenheit a Kelvin")
    );

    private String categoriaActual;
    private BiConsumer<String, Double> accionConvertir;
    private Runnable accionVolver;

    public PanelConversion() {
        initComponents();
        configurarVista();
        conectarEventos();
    }

    private void configurarVista() {
        setBackground(Paleta.GRIS_FONDO);

        lblEncabezado.setForeground(Paleta.AZUL);
        lblEncabezado.setFont(Paleta.TITULO);

        btnConvertir.setBackground(Paleta.AZUL);
        btnConvertir.setForeground(Color.WHITE);
        btnConvertir.setOpaque(true);
        btnConvertir.setBorderPainted(false);
        btnConvertir.setContentAreaFilled(true);

        lblResultado.setOpaque(true);
        lblResultado.setBackground(Color.WHITE);
    }

    private void conectarEventos() {
        btnConvertir.addActionListener(e -> dispararConvertir());
        campoValor.addActionListener(e -> dispararConvertir());
        btnVolver.addActionListener(e -> {
            if (accionVolver != null) accionVolver.run();
        });
    }

    private void dispararConvertir() {
        Opcion seleccion = (Opcion) comboOperacion.getSelectedItem();
        if (seleccion == null) return;
        String texto = campoValor.getText().trim().replace(',', '.');
        try {
            double valor = Double.parseDouble(texto);
            if (accionConvertir != null) accionConvertir.accept(seleccion.clave, valor);
        } catch (NumberFormatException ex) {
            mostrarResultado(Resultado.error("Ingresa un número válido (ej: 12.5)"));
        }
    }

    // ========= API publica consumida por ControladorEscritorio =========

    @SuppressWarnings("unchecked")
    public void setCategoria(String categoria) {
        this.categoriaActual = categoria;
        List<Opcion> opciones;
        String titulo;
        switch (categoria) {
            case "longitud":
                opciones = OPCIONES_LONGITUD;
                titulo = "Conversiones de Longitud";
                break;
            case "masa":
                opciones = OPCIONES_MASA;
                titulo = "Conversiones de Masa";
                break;
            case "temperatura":
                opciones = OPCIONES_TEMPERATURA;
                titulo = "Conversiones de Temperatura";
                break;
            default:
                opciones = List.of();
                titulo = "Conversiones";
        }
        lblEncabezado.setText(titulo);
        comboOperacion.setModel(new DefaultComboBoxModel(opciones.toArray()));
        campoValor.setText("");
        lblResultado.setText(" ");
        lblResultado.setBackground(Color.WHITE);
    }

    public String getCategoria() {
        return categoriaActual;
    }

    public void mostrarResultado(Resultado resultado) {
        if (resultado == null) return;
        if (resultado.isExito()) {
            lblResultado.setText("Resultado: " + resultado.getValor());
            lblResultado.setBackground(Paleta.VERDE_EXITO_BG);
            lblResultado.setForeground(Paleta.VERDE_EXITO_FG);
        } else {
            lblResultado.setText(resultado.getMensaje());
            lblResultado.setBackground(Paleta.ROJO_ERROR_BG);
            lblResultado.setForeground(Paleta.ROJO_ERROR_FG);
        }
    }

    public void setBotonHabilitado(boolean habilitado) {
        btnConvertir.setEnabled(habilitado);
        btnConvertir.setText(habilitado ? "Convertir" : "Convirtiendo...");
    }

    public void setOnConvertir(BiConsumer<String, Double> accion) {
        this.accionConvertir = accion;
    }

    public void setOnVolver(Runnable accion) {
        this.accionVolver = accion;
    }

    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    @SuppressWarnings("unchecked")
    private void initComponents() {

        lblEncabezado = new javax.swing.JLabel();
        lblConversion = new javax.swing.JLabel();
        comboOperacion = new javax.swing.JComboBox<>();
        lblValor = new javax.swing.JLabel();
        campoValor = new javax.swing.JTextField();
        btnConvertir = new javax.swing.JButton();
        lblResultado = new javax.swing.JLabel();
        btnVolver = new javax.swing.JButton();

        lblEncabezado.setFont(new java.awt.Font("SansSerif", 1, 20)); // NOI18N
        lblEncabezado.setText("Conversiones");

        lblConversion.setFont(new java.awt.Font("SansSerif", 1, 13)); // NOI18N
        lblConversion.setText("Conversión:");

        lblValor.setFont(new java.awt.Font("SansSerif", 1, 13)); // NOI18N
        lblValor.setText("Valor:");

        btnConvertir.setFont(new java.awt.Font("SansSerif", 1, 13)); // NOI18N
        btnConvertir.setText("Convertir");
        btnConvertir.setFocusPainted(false);

        lblResultado.setHorizontalAlignment(javax.swing.SwingConstants.CENTER);
        lblResultado.setText(" ");
        lblResultado.setBorder(javax.swing.BorderFactory.createEmptyBorder(10, 12, 10, 12));

        btnVolver.setText("Volver al Menú");
        btnVolver.setFocusPainted(false);

        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(this);
        this.setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addGap(40, 40, 40)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(lblEncabezado)
                    .addComponent(lblConversion)
                    .addComponent(comboOperacion, 0, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(lblValor)
                    .addComponent(campoValor)
                    .addComponent(btnConvertir, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(lblResultado, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(btnVolver, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
                .addGap(40, 40, 40))
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addGap(24, 24, 24)
                .addComponent(lblEncabezado)
                .addGap(16, 16, 16)
                .addComponent(lblConversion)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addComponent(comboOperacion, javax.swing.GroupLayout.PREFERRED_SIZE, 30, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(10, 10, 10)
                .addComponent(lblValor)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addComponent(campoValor, javax.swing.GroupLayout.PREFERRED_SIZE, 30, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(14, 14, 14)
                .addComponent(btnConvertir, javax.swing.GroupLayout.PREFERRED_SIZE, 36, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(12, 12, 12)
                .addComponent(lblResultado, javax.swing.GroupLayout.PREFERRED_SIZE, 40, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(10, 10, 10)
                .addComponent(btnVolver, javax.swing.GroupLayout.PREFERRED_SIZE, 30, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );
    }// </editor-fold>//GEN-END:initComponents


    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JButton btnConvertir;
    private javax.swing.JButton btnVolver;
    private javax.swing.JTextField campoValor;
    private javax.swing.JComboBox<Object> comboOperacion;
    private javax.swing.JLabel lblConversion;
    private javax.swing.JLabel lblEncabezado;
    private javax.swing.JLabel lblResultado;
    private javax.swing.JLabel lblValor;
    // End of variables declaration//GEN-END:variables
}

File: SIN BDD/SOAP_DOTNET_CONUNI_GR06/05. CLIESCRITORIO/cliente_escritorio_soap_dotnet_conuni_gr06/src/Vista/PanelConversion.cs
using Ec.Edu.Monster.Controlador;
using System.Drawing;
using System.Globalization;
using System.IO;
using System.Windows.Forms;

namespace Ec.Edu.Monster.Vista;

public partial class PanelConversion : UserControl
{
    public event Action? VolverSolicitado;

    private readonly ControladorEscritorio controlador;
    private readonly string categoria;

    public PanelConversion(ControladorEscritorio controlador, string categoria)
    {
        this.controlador = controlador;
        this.categoria = categoria;
        InitializeComponent();
        ConstruirInterfaz();
    }

    private void ConstruirInterfaz()
    {
        BackColor = Color.White;
        panelEncabezado.BackColor = Paleta.AZUL;
        btnVolver.BackColor = Paleta.AMARILLO;
        btnVolver.ForeColor = Paleta.AZUL;
        btnVolver.FlatStyle = FlatStyle.Flat;
        btnVolver.FlatAppearance.BorderSize = 0;
        lblTituloCategoria.Font = Paleta.TITULO;
        lblTituloCategoria.ForeColor = Color.White;
        lblEncabezado.Font = new Font("SansSerif", 20, FontStyle.Bold);
        lblEncabezado.ForeColor = Paleta.AZUL;
        lblDescripcion.Font = Paleta.SUBTITULO;
        lblDescripcion.ForeColor = Paleta.TEXTO_SUAVE;
        btnConvertir.BackColor = Paleta.AZUL;
        btnConvertir.ForeColor = Color.White;
        btnConvertir.FlatStyle = FlatStyle.Flat;
        btnConvertir.FlatAppearance.BorderSize = 0;
        lblResultado.BackColor = Color.White;

        btnVolver.Click += (_, _) => VolverSolicitado?.Invoke();
        lblTituloCategoria.Text = TituloCategoria();
        lblEncabezado.Text = $"Conversiones de {TituloCategoria()}";
        lblDescripcion.Text = DescripcionCategoria();
        campoValor.KeyDown += (_, e) =>
        {
            if (e.KeyCode == Keys.Enter)
            {
                Convertir();
                e.SuppressKeyPress = true;
            }
        };
        btnConvertir.Click += (_, _) => Convertir();
        CargarOpciones();
    }

    private void CargarOpciones()
    {
        comboOperacion.Items.Clear();
        var opciones = categoria switch
        {
            "longitud" => new[]
            {
                ("metrosAPies", "Metros a Pies"),
                ("kilometrosAMillas", "Kilometros a Millas"),
                ("centimetrosAPulgadas", "Centimetros a Pulgadas"),
                ("yardasAMetros", "Yardas a Metros"),
                ("milimetrosAPulgadas", "Milimetros a Pulgadas")
            },
            "masa" => new[]
            {
                ("kilogramosALibras", "Kilogramos a Libras"),
                ("gramosAOnzas", "Gramos a Onzas"),
                ("toneladasAKilogramos", "Toneladas a Kilogramos"),
                ("librasAOnzas", "Libras a Onzas"),
                ("miligramosAGramos", "Miligramos a Gramos")
            },
            _ => new[]
            {
                ("celsiusAFahrenheit", "Celsius a Fahrenheit"),
                ("fahrenheitACelsius", "Fahrenheit a Celsius"),
                ("celsiusAKelvin", "Celsius a Kelvin"),
                ("kelvinACelsius", "Kelvin a Celsius"),
                ("fahrenheitAKelvin", "Fahrenheit a Kelvin")
            }
        };

        foreach (var opcion in opciones)
        {
            comboOperacion.Items.Add(new OpcionConversion(opcion.Item1, opcion.Item2));
        }

        comboOperacion.DisplayMember = nameof(OpcionConversion.Etiqueta);
        comboOperacion.ValueMember = nameof(OpcionConversion.Clave);
        comboOperacion.SelectedIndex = 0;
    }

    private void Convertir()
    {
        if (!double.TryParse(campoValor.Text.Replace(',', '.'), NumberStyles.Any, CultureInfo.InvariantCulture, out var valor))
        {
            lblResultado.Text = "Ingresa un numero valido";
            lblResultado.ForeColor = Color.FromArgb(0xA1, 0x00, 0x00);
            return;
        }

        if (comboOperacion.SelectedItem is not OpcionConversion opcion)
        {
            return;
        }

        var resultado = categoria switch
        {
            "longitud" => controlador.ConvertirLongitud(opcion.Clave, valor),
            "masa" => controlador.ConvertirMasa(opcion.Clave, valor),
            _ => controlador.ConvertirTemperatura(opcion.Clave, valor)
        };

        if (resultado.Exito)
        {
            lblResultado.Text = $"Resultado: {resultado.Valor:0.000}";
            lblResultado.ForeColor = Color.FromArgb(0x00, 0x6B, 0x00);
            lblResultado.BackColor = Color.FromArgb(0xE2, 0xFD, 0xE2);
        }
        else
        {
            lblResultado.Text = resultado.Mensaje;
            lblResultado.ForeColor = Color.FromArgb(0xA1, 0x00, 0x00);
            lblResultado.BackColor = Color.FromArgb(0xFD, 0xE2, 0xE2);
        }
    }

    private string TituloCategoria() => categoria switch
    {
        "longitud" => "Longitud",
        "masa" => "Masa",
        "temperatura" => "Temperatura",
        _ => categoria
    };

    private string DescripcionCategoria() => categoria switch
    {
        "longitud" => "Convierte entre metros, pies, kilometros, millas, pulgadas y mas.",
        "masa" => "Convierte entre kilogramos, libras, gramos, onzas, toneladas y mas.",
        _ => "Convierte entre Celsius, Fahrenheit y Kelvin."
    };

    private static string Ruta(string archivo)
    {
        var rutaPrincipal = Path.Combine(AppContext.BaseDirectory, "src", "img", archivo);
        if (File.Exists(rutaPrincipal))
        {
            return rutaPrincipal;
        }

        return Path.Combine(AppContext.BaseDirectory, "img", archivo);
    }

    private sealed class OpcionConversion
    {
        public OpcionConversion(string clave, string etiqueta)
        {
            Clave = clave;
            Etiqueta = etiqueta;
        }

        public string Clave { get; }

        public string Etiqueta { get; }

        public override string ToString() => Etiqueta;
    }

    private void lblEncabezado_Click(object sender, EventArgs e)
    {

    }

    private void lblDescripcion_Click(object sender, EventArgs e)
    {

    }
}
